File: interview/coinChange.go
package interview

import "math"

// min[i] = min[i - c] + 1, c => [coins]
func coinChange(coins []int, amount int) int {
	if amount < 1 && len(coins) < 1 {
		return -1
	}
	memo := make([]int, amount+1)
	for i := 1; i <= amount; i++ {
		memo[i] = math.MaxInt32
		for _, c := range coins {
			if i < c {
				break
			}
			if memo[i] > memo[i-c]+1 {
				memo[i] = memo[i-c] + 1
			}
		}
	}
	if memo[amount] == math.MaxInt32 {
		return -1
	}
	return memo[amount]
}

File: algo-java/src/专项突破/动态规划/背包问题/CoinChange.java
package 专项突破.动态规划.背包问题;

/**
 * @author Zexho
 * @date 2021/11/30 4:09 下午
 */
public class CoinChange {
    public int coinChange(int[] coins, int amount) {
        int[] arr = new int[amount + 1];
        for (int i = 1; i < arr.length; i++) {
            arr[i] = amount + 1;
            for (int coin : coins) {
                if (i >= coin) {
                    arr[i] = Math.min(arr[i], arr[i - coin] + 1);
                }
            }
        }
        return arr[amount] > amount ? -1 : arr[amount];
    }
}
